### PR TITLE
Refactor/tao internal activity/model

### PR DESCRIPTION
**Pull request by sirpa44**

### CODE REVIEW

File: Src/App/Router.php [hunk 1]
```diff
@@ -8,30 +8,25 @@ class Router
...
      * lead to the controller
      * @param $parameters
-     * @return mixed
+     * @return array
      * @throws \Exception
      */
     public function route($parameters)
     {
-
+        $controllerClassName = ucfirst($parameters['controller']) . 'Controller';
```
Comment (siwane) [inline Src/App/Router.php]: `$parameters['controller']` may not exist, you should test it before

File: Src/App/Router.php [hunk 2]
```diff
@@ -8,30 +8,25 @@ class Router
...
-            $controllerClassName = ucfirst($parameters['controller']) . 'Controller';
             $controllerName = '\Aot\Controller\\' . $controllerClassName;
             $method = $parameters['method'];
         } else {
             throw new \Exception("method or controller invalid");
         }
         if (class_exists($controllerName)) {
             $controller = new $controllerName();
+            unset($parameters['method'], $parameters['controller']);
+            return $controller->action($parameters, $method);
```
Comment (siwane) [inline Src/App/Router.php]: How do know `action` exists in the controller?

File: Src/Model/Adapter/AdapterInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<?php
+namespace Aot\Model\Adapter;
+
+interface AdapterInterface
```
Comment (siwane) [inline Src/Model/Adapter/AdapterInterface.php:4]: :clap: 
I'm aplausing, ur first interface! However, an interface has to constraint a class and must be correctly documented

File: Src/Model/Factory/FormatFactory.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?php
+namespace Aot\Model\Factory;
+
+class FormatFactory
+{
+    private $format = ['csv','json'];
+    private $path = 'Aot\Model\Adapter\\';
```
Comment (siwane) [inline Src/Model/Factory/FormatFactory.php]: `\` or `\\` ?

File: Src/Model/Factory/FormatFactory.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    /**
+     * create an instance of adapter
+     * @param $format
+     * @return instance
+     * @throws \Exception
+     */
+    public function getFormatInstance($format)
+    {
+        if (in_array($format, $this->format)) {
+            $classPath = $this->path . ucfirst($format);
```
Comment (siwane) [inline Src/Model/Factory/FormatFactory.php]: Several things here,  good concept! 
However the class may not exist and even if it exists u dont know if it implements `AdapterInterface` or not. You can check `is_a` function in php manual

**Review by siwane — CHANGES_REQUESTED** [review]
This a great start for interface! I'm proud of u! 
I have left several comments and I'm letting u throwing an eye on it. 
In all cases, u can brainstorm about the idea of the interface method called is coming from http parameters. I'm hoping the `delete` method will not appear in the interface

File: Src/Model/Factory/FormatFactory.php (hunk 4 above)
Comment (sirpa44) [reply Src/Model/Factory/FormatFactory.php]: if i put only one \ the ' is disable

File: Src/App/Router.php (hunk 1 above)
Comment (sirpa44) [reply Src/App/Router.php]: done

File: Src/Model/Factory/FormatFactory.php (hunk 4 above)
Comment (siwane) [reply Src/Model/Factory/FormatFactory.php]: You mean `\M` and `\A` are not escaped? If you want `\` as a string, always escape it

Comment (sirpa44) [reply Src/Model/Factory/FormatFactory.php]: i mean 'l\'ecole\' enseigne la vie' where stop the string?

Comment (sirpa44) [reply Src/Model/Factory/FormatFactory.php]: to escape a special character i use  \  . and  \  is a special character if he is use to escape a special character. and M and A are not special character then  \  is not a special character.

File: Src/App/Router.php (hunk 2 above)
Comment (sirpa44) [reply Src/App/Router.php]: action is hard coded should i control it anyway?

File: Src/Model/Factory/FormatFactory.php (hunk 4 above)
Comment (sirpa44) [reply Src/Model/Factory/FormatFactory.php]: http://php.net/language.types.string

File: Src/Model/Factory/FormatFactory.php (hunk 5 above)
Comment (sirpa44) [reply Src/Model/Factory/FormatFactory.php]: i don't understand !

File: Src/Model/Adapter/AdapterInterface.php (hunk 3 above)
Comment (sirpa44) [reply Src/Model/Adapter/AdapterInterface.php:4]: done

File: Src/Model/Adapter/Adapter.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<?php
+namespace Aot\Model\Adapter;
+
+abstract class Adapter
```
Comment (siwane) [inline Src/Model/Adapter/Adapter.php]: You must have an interface on top of the abstract class. It allows to implement the interface without the logic from abstract. This implementation is specific to file

File: Src/App/Router.php [hunk 7]
```diff
@@ -4,35 +4,30 @@
...
             throw new \Exception("controller invalid");
         }
-        if (is_callable($method, $controllerClassName)) {
-            unset($parameters['method'], $parameters['controller']);
-            return $controller->$method($parameters);
-            } else {
-            throw new \Exception("method invalid");
+        $controller = new $controllerName();
+        unset($parameters['method'], $parameters['controller']);
+        if (!is_a($controller, $this->interfaceControllerPath)) {
```
Comment (siwane) [inline Src/App/Router.php]: with a `true` as third parameter you can remove the `class_exists` check

Comment (sirpa44) [reply Src/App/Router.php]: done

**Pull request merged**